### DISCUSSION TITLE
How to run HAN on link prediction tasks?

### DISCUSSION THREAD

**Discussion by xubingze** [pyg-team/pytorch_geometric#8055] [Q&A] [answered]
Hi everyone, I am new to this field, and I am trying to implement HAN on my own dataset without any labels. I noticed this official PyG [example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/han_imdb.py), but I have two questions regarding the implementation. Firstly, my dataset has four different kinds of relation between same two type of nodes, meaning that: Node type A - Link type 1- Node type A, Node type A - Link type 2- Node type A, Node type A - Link type 3- Node type A, and Node type A - Link type 4- Node type A. In this case, how shall I edit the methpaths? Secondly, can anyone give me an example on how to adjust the codes to the unsupervised link prediction tasks? Thank you all very much!

**Comment by rusty1s** [comment]
Hey, sorry for late reply. In general, it is not straightforward to apply metapaths in your case, and I don't think it is particularly useful if you only have a single node type. In this case, using something like `RGCNConv` sounds better.

There also exists an example for link prediction via R-GCN: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/rgcn_link_pred.py

**Reply by xubingze** [reply]
Thank you very much for your reply. I forgot to mention that actually we have other node types, Node Type A - Link Type 5 - Node type B and Node Type A - Link Type 6 - Node Type C. I wonder can I still apply "RGCNConv" in this case.

**Reply by rusty1s** [reply, marked as answer]
I see, in this case R-GCN is still applicable but need some more work to make it work. In particular, you need to ensure that your input features are projected into a shared embedding space. This could look something like this:
```python
def forward(self, x_dict: Dict[str, Tensor], edge_index, edge_type):
    x_dict = self.hetero_linear(data.x_dict)
    x = torch.cat([x for x in x_dict.values()], dim=0)
    x = self.rgcn_conv(x, edge_index, edge_type)
```
where `x_dict` is coming from `HeteroData.x_dict` and `edge_index` and `edge_type` are coming from `HeteroData.to_homogeneous()`.
